portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -59,40 +62,38 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String nickname = getParameter(request, "nickname", "");
+    String nickname = getParameter(request, "nickname").orElse("");
     String id = userService.getCurrentUser().getUserId();
 
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    Entity entity = new Entity("UserInfo");
-    entity.setProperty("id", id);
-    entity.setProperty("nickname", nickname);
-    
-    datastore.put(entity);
+
+    Entity userInfoEntity = getUserinfoEntity(id);
+
+    if (userInfoEntity == null) {
+      userInfoEntity = new Entity("UserInfo");
+      userInfoEntity.setProperty("id", id);
+    }
+
+    userInfoEntity.setProperty("nickname", nickname);
+    datastore.put(userInfoEntity);
 
     response.sendRedirect("/index.html");
   }
 
-  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
-    String value = request.getParameter(name);
-    if (value == null) {
-      return defaultValue;
-    }
-    return value;
+  private Optional<String> getParameter(HttpServletRequest request, String name) {
+    return Optional.ofNullable(request.getParameter(name));
   }
 
   /**
-   * Returns the nickname of the user with id, or empty String if the user has not set a nickname.
+   * Returns the UserInfo entity with user id.
+   * Given id is not of UserInfo kind but a field of that kind.
    */
-  private String getUserNickname(String id) {
+  private Entity getUserinfoEntity(String id) {
```

Comment:
Return Optional<Entity> instead of nullable Entity. Can use `Optional.ofNullable(entity);`